In [2]:
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## 1.4

### a)

Nodes $x_0 = 0,\ x_1 = 1,\ x_2 = 4$, values $f_0 = 0,\ f_1 = 2,\ f_2 = 8$.

$$\ell_0(x) = \frac{(x-1)(x-4)}{(0-1)(0-4)} = \frac{(x-1)(x-4)}{4}$$

$$\ell_1(x) = \frac{(x-0)(x-4)}{(1-0)(1-4)} = -\frac{x(x-4)}{3}$$

$$\ell_2(x) = \frac{(x-0)(x-1)}{(4-0)(4-1)} = \frac{x(x-1)}{12}$$

$$p(x) = \sum_{i=0}^{2} f_i\,\ell_i(x) = 0\cdot\frac{(x-1)(x-4)}{4} + 2\cdot\left(-\frac{x(x-4)}{3}\right) + 8\cdot\frac{x(x-1)}{12}$$

$$= -\frac{2x(x-4)}{3} + \frac{2x(x-1)}{3} = \frac{2x}{3}\big(-(x-4) + (x-1)\big) = \frac{2x}{3}\cdot 3 = 2x$$

### b)

![4b](drawings/4b.png)

### c)

In [6]:
x = np.array([0, 1, 4])
f = np.array([0, 2, 8])
coeffs = np.polyfit(x, f, 2)
print(coeffs)
np.polyval(coeffs, 2)

[7.48707318e-16 2.00000000e+00 7.13708072e-16]


np.float64(4.000000000000003)

The coefficients from `polyfit` are approximately equal to the ones computed by hand,
but returned in reverse order $[a_2, a_1, a_0]$. Due to machine rounding errors
$a_2 \approx 7.5\cdot10^{-16}$ and $a_0 \approx 7.1\cdot10^{-16}$ instead of exactly 0.

`polyval` computed the correct value $p(2) = 4$ up to rounding ($4.000000000000003$).